### PR TITLE
Client and Models implementation of access management for clusters, cluster groups, workspace, namespace and org

### PR DESCRIPTION
**Pull request by Vasundharashukla**
Signed-off-by: Vasundhara Shukla <vasundharas@vmware.com>

1. **What this PR does / why we need it**: Client and Models needed to implement Role Bindings support in provider

2. **Which issue(s) this PR fixes**

        (optional, in `fixes #<issue number>` format, will close the issue(s) when PR gets merged):

        Fixes #

3. **Additional information**
Task: [OLYMP-30505](https://jira.eng.vmware.com/browse/OLYMP-30505)

4. **Special notes for your reviewer**

### CODE REVIEW

**Comment by shreyassreenivas** [comment]
@Vasundharashukla , can you please update MR template with the right details

File: docs/index.md [hunk 1]
```diff
@@ -56,9 +56,9 @@ provider "tanzu-mission-control" {
 
 ### Required
 
-- **endpoint** (String)
-- **vmw_cloud_api_token** (String, Sensitive)
+- `endpoint` (String)
```
Comment (shreyassreenivas) [inline docs/index.md:59]: Why is this change needed ?

File: internal/client/cluster/cluster_resource.go [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright © 2021 VMware, Inc. All Rights Reserved.
+Copyright © 2022 VMware, Inc. All Rights Reserved.
```
Comment (shreyassreenivas) [inline internal/client/cluster/cluster_resource.go:2]: This should not be updated

File: internal/client/cluster/iam_policy/iam_resource.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+type ClientService interface {
+	ManageV1alpha1ClusterIAMPolicyGet(fn *clustermodel.VmwareTanzuManageV1alpha1ClusterFullName) (*clusteriammodel.VmwareTanzuManageV1alpha1ClusterGetClusterIAMPolicyResponse, error)
+
+	ManageV1alpha1ClusterIAMPolicyPatch(request *clusteriammodel.VmwareTanzuManageV1alpha1ClusterPatchClusterIAMPolicyRequest) (*clusteriammodel.VmwareTanzuManageV1alpha1ClusterPatchClusterIAMPolicyResponse, error)
+
+	ManageV1alpha1ClusterIAMPolicyUpdate(fn *clustermodel.VmwareTanzuManageV1alpha1ClusterFullName, request *iammodel.VmwareTanzuCoreV1alpha1PolicyIAMPolicy) (*clusteriammodel.VmwareTanzuManageV1alpha1ClusterUpdateClusterIAMPolicyResponse, error)
+}
+
+/*
+  ManageV1alpha1ClusterIAMPolicyGet gets policy for a cluster.
```
Comment (shreyassreenivas) [inline internal/client/cluster/iam_policy/iam_resource.go:40]: gets all iam policies scoped to a cluster. 

On the similar lines can you do the changes  for the below function comments as well

File: internal/client/cluster/iam_policy/iam_resource.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+	if fn.ManagementClusterName != "" {
+		queryParams["fullName.managementClusterName"] = []string{fn.ManagementClusterName}
+	}
+
+	if fn.ProvisionerName != "" {
+		queryParams["fullName.provisionerName"] = []string{fn.ProvisionerName}
+	}
+
+	requestURL := fmt.Sprintf("%s/%s?%s", "v1alpha1/clusters:iam", fn.Name, queryParams.Encode())
```
Comment (shreyassreenivas) [inline internal/client/cluster/iam_policy/iam_resource.go:54]: "v1alpha1/clusters:iam" this can be made into a constant and referred in all the places

File: internal/client/cluster/iam_policy/iam_resource.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+/*
+  ManageV1alpha1ClusterIAMPolicyPatch patches a cluster policy.
+*/
+
+func (c *Client) ManageV1alpha1ClusterIAMPolicyPatch(request *clusteriammodel.VmwareTanzuManageV1alpha1ClusterPatchClusterIAMPolicyRequest) (*clusteriammodel.VmwareTanzuManageV1alpha1ClusterPatchClusterIAMPolicyResponse, error) {
+	response := &clusteriammodel.VmwareTanzuManageV1alpha1ClusterPatchClusterIAMPolicyResponse{}
+	requestURL := fmt.Sprintf("%s/%s", "v1alpha1/clusters:iam", request.FullName.Name)
+	err := c.Patch(requestURL, request, response)
```
Comment (shreyassreenivas) [inline internal/client/cluster/iam_policy/iam_resource.go:68]: Patch has not been tested till now, so please test this client implementation

File: internal/client/clustergroup/cluster_group_resource.go [hunk 6]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright © 2021 VMware, Inc. All Rights Reserved.
+Copyright © 2022 VMware, Inc. All Rights Reserved.
```
Comment (shreyassreenivas) [inline internal/client/clustergroup/cluster_group_resource.go:2]: Why is this being updated

File: internal/client/clustergroup/iam_policy/iam_resource.go [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+/*
+  Client for cluster group IAM policy resource service API.
+*/
+
+type Client struct {
+	*transport.Client
+}
+
+// ClientService is the interface for Client methods.
```
Comment (shreyassreenivas) [inline internal/client/clustergroup/iam_policy/iam_resource.go:30]: ClientService is the interface for ManageV1alpha1ClusterGroupIAMPolicy Client methods.

File: internal/client/clustergroup/iam_policy/iam_resource.go [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+	ManageV1alpha1ClusterGroupIAMPolicyUpdate(fn *clustergroupmodel.VmwareTanzuManageV1alpha1ClustergroupFullName, request *iammodel.VmwareTanzuCoreV1alpha1PolicyIAMPolicy) (*clustergroupiammodel.VmwareTanzuManageV1alpha1ClustergroupUpdateClusterGroupIAMPolicyResponse, error)
+}
+
+/*
+  ManageV1alpha1ClusterGroupIAMPolicyGet gets policy for a cluster group.
+*/
+
+func (c *Client) ManageV1alpha1ClusterGroupIAMPolicyGet(fn *clustergroupmodel.VmwareTanzuManageV1alpha1ClustergroupFullName) (*clustergroupiammodel.VmwareTanzuManageV1alpha1ClustergroupGetClusterGroupIAMPolicyResponse, error) {
+	requestURL := fmt.Sprintf("%s/%s", "v1alpha1/clustergroups:iam", fn.Name)
```
Comment (shreyassreenivas) [inline internal/client/clustergroup/iam_policy/iam_resource.go:44]: make it a constant "v1alpha1/clustergroups:iam" and similar changes in other places as well

File: internal/client/http_client.go [hunk 9]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright © 2021 VMware, Inc. All Rights Reserved.
+Copyright © 2022 VMware, Inc. All Rights Reserved.
```
Comment (shreyassreenivas) [inline internal/client/http_client.go:2]: Please revert this change

File: internal/client/transport/methods.go [hunk 10]
```diff
@@ -33,6 +33,10 @@ func (c *Client) Update(url string, request Request, response Response) error {
 	return c.invokeAction(http.MethodPut, url, request, response)
 }
 
+func (c *Client) Patch(url string, request Request, response Response) error {
```
Comment (shreyassreenivas) [inline internal/client/transport/methods.go:36]: Is this tested ?

File: resource_templates/iam_policy.tf [hunk 11]
```diff
@@ -0,0 +1,29 @@
+// Update Tanzu Mission Control iam policy entry
+resource "tanzu-mission-control_iam" "update_iam" {
```
Comment (shreyassreenivas) [inline resource_templates/iam_policy.tf:2]: please add this file when the resource is implemented and tested. Then it will reflect the latest and working example

File: internal/client/cluster/iam_policy/iam_resource.go (hunk 5 above)
Comment (Vasundharashukla) [reply internal/client/cluster/iam_policy/iam_resource.go:68]: I created the role binding using path operation, so it is tested.

File: internal/client/organization/iam_policy/iam_resource.go [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+type ClientService interface {
+	ManageV1alpha1OrganizationIAMPolicyGet(fn *organizationmodel.VmwareTanzuManageV1alpha1OrganizationFullName) (*organizationiammodel.VmwareTanzuManageV1alpha1OrganizationGetOrganizationIAMPolicyResponse, error)
+
+	ManageV1alpha1OrganizationIAMPolicyPatch(request *organizationiammodel.VmwareTanzuManageV1alpha1OrganizationPatchOrganizationIAMPolicyRequest) (*organizationiammodel.VmwareTanzuManageV1alpha1OrganizationPatchOrganizationIAMPolicyResponse, error)
+
+	ManageV1alpha1OrganizationIAMPolicyUpdate(fn *organizationmodel.VmwareTanzuManageV1alpha1OrganizationFullName, request *iammodel.VmwareTanzuCoreV1alpha1PolicyIAMPolicy) (*organizationiammodel.VmwareTanzuManageV1alpha1OrganizationUpdateOrganizationIAMPolicyResponse, error)
+}
+
+/*
+  ManageV1alpha1OrganizationIAMPolicyGet gets all iam policies scoped to a organization.
```
Comment (shreyassreenivas) [inline internal/client/organization/iam_policy/iam_resource.go:43]: a to an

**Review by shreyassreenivas — CHANGES_REQUESTED** [review]

File: docs/index.md (hunk 1 above)
Comment (Vasundharashukla) [reply docs/index.md]: That was taken in as we hadn't run the documentation, the change was as per some old version. I have reverted it.

File: internal/client/cluster/iam_policy/iam_resource.go (hunk 3 above)
Comment (Vasundharashukla) [reply internal/client/cluster/iam_policy/iam_resource.go]: Done at all places

File: internal/client/cluster/iam_policy/iam_resource.go (hunk 4 above)
Comment (Vasundharashukla) [reply internal/client/cluster/iam_policy/iam_resource.go]: Made constants as `paths` and replaced in all client files.

File: internal/client/clustergroup/cluster_group_resource.go (hunk 6 above)
Comment (Vasundharashukla) [reply internal/client/clustergroup/cluster_group_resource.go]: Missed to revert this while moving iam policy client to separate directories. Changed

File: resource_templates/iam_policy.tf (hunk 11 above)
Comment (Vasundharashukla) [reply resource_templates/iam_policy.tf]: Sure, removed it from the PR

**Review by shreyassreenivas — APPROVED** [review]

**Comment by ishangupta-ds** [comment]
LGTM!

**Review by ishangupta-ds — APPROVED** [review]
